apis/v1alpha2/gatewayclass_types.go

```diff
@@ -0,0 +1,205 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api,scope=Cluster,shortName=gc
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Controller",type=string,JSONPath=`.spec.controller`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:printcolumn:name="Description",type=string,JSONPath=`.spec.description`,priority=1
+
+// GatewayClass describes a class of Gateways available to the user
+// for creating Gateway resources.
+//
+// GatewayClass is a Cluster level resource.
+type GatewayClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of GatewayClass.
+	Spec GatewayClassSpec `json:"spec"`
+
+	// Status defines the current state of GatewayClass.
+	//
+	// +kubebuilder:default={conditions: {{type: "Admitted", status: "False", message: "Waiting for controller", reason: "Waiting", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status GatewayClassStatus `json:"status,omitempty"`
+}
+
+// GatewayClassSpec reflects the configuration of a class of Gateways.
+type GatewayClassSpec struct {
+	// Controller is a domain/path string that indicates the
+	// controller that is managing Gateways of this class.
+	//
+	// Example: "acme.io/gateway-controller".
+	//
+	// This field is not mutable and cannot be empty.
+	//
+	// The format of this field is DOMAIN "/" PATH, where DOMAIN is a valid
+	// Kubernetes name (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names)
+	// and PATH is a valid HTTP path as defined by RFC 3986.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MaxLength=253
+	Controller string `json:"controller"`
+
+	// ParametersRef is a reference to a resource that contains the configuration
+	// parameters corresponding to the GatewayClass. This is optional if the
+	// controller does not require any additional configuration.
+	//
+	// ParametersRef can reference a standard Kubernetes resource, i.e. ConfigMap,
+	// or an implementation-specific custom resource. The resource can be
+	// cluster-scoped or namespace-scoped.
+	//
+	// If the referent cannot be found, the GatewayClass's "InvalidParameters"
+	// status condition will be true.
+	//
+	// Support: Custom
+	//
+	// +optional
+	ParametersRef *ParametersReference `json:"parametersRef,omitempty"`
```

Comment:
So parameter ref is something that the controller can only interpret, right? Whatever it is, even if it was one of the core objects (secrets, configmap). if so then why not go with `map[string]string` (validated for lengths) then let the controller decide if they are valid/not valid or not.

I am also concerned with updated "parameterRef`. e.g., i changed the GWClass spec.ParamterRef. How can the controller patch status to tell `your updated parameter ref` is borked and right now all gateways that uses this class is broken.

Reply:
Yeah, this is an identical pattern to IngressClass and it has similar issues. StorageClass took the `map[string]string` approach, but it also predates CRDs. The advantage of CRD based params is that there's at least some form of validation involved if you have a typo in a field name or an invalid value. 

I agree with your concerns around param lifecycle. We need to discuss this more. I know at least some implementations are using the GatewayClass and associated params as a template for a Gateway as it is created, and ignoring any future changes to the class resource or params for existing Gateways. That feels like the safest way to avoid this problem, but that does require controllers to have a clear way to communicate the configuration a Gateway is being implemented with.

Reply:
Discussed at community meeting, will add guidance suggesting implementers should use template behavior. If not using this template based approach it must be clearly documented.

Reply:
Will be fixed by #839